torch/_library/vmap.py

```diff
@@ -0,0 +1,45 @@
+# mypy: allow-untyped-defs
+import dataclasses
+from typing import Callable, Tuple, Optional, Protocol
+
+from .. import _ops, autograd
+
+
+class InfoProtocol(Protocol):
+    _vmap_fn: Optional[Callable]
+    _setup_context_fn: Optional[Callable]
+
+@dataclasses.dataclass
+class Info:
+    _vmap_fn: Optional[Callable]
+    _setup_context_fn: Optional[Callable]
+
+def make_vmap_impl(op: _ops.OpOverload, info: InfoProtocol) -> Callable:
+    name: str = f"GeneratedVmapFor_{op._namespace}_{op._opname}_{op._overloadname}"
+
+    def vmap(ctx, in_dims: Tuple[Optional[int]], *args, **kwargs):
+        if info._vmap_fn:
+            result = info._vmap_fn(ctx, in_dims, *args, **kwargs)
+
+            return result
+
+        raise RuntimeError(
+            f"Trying to vmap through {op} but no vmap "
+            f"formula was registered. "
+            f"Please use register_vmap to add one."
+        )
+
+    Generated = type(
+        name,
+        (autograd.Function,),
+        {
+            "setup_context": staticmethod(info._setup_context_fn),
+            "vmap": staticmethod(vmap),
+        },
+    )
+
+    def vmap_impl(keyset, *args, **kwargs):
+        return Generated.apply(*args, **kwargs)  # type: ignore[attr-defined]
```

Comment:
Hmm, I know I suggested this, but I don't think this is the right approach. The user should not need to give us a setup_context staticmethod.

We basically want the vmap_impl to just be this piece of code https://github.com/pytorch/pytorch/blob/febadda107e5bfd78d32a276000a313b3b10f41b/torch/_functorch/autograd_function.py#L305-L334 that we can share between autograd.Function and this. (We should drop the user's vmap function into where the call to `autograd_function.vmap` is.)

Reply:
I changed to this approach now. 

I'm not sure what should we do for the lines below in the custom vmap case if we don't have a autograd_function?
```
 if pytree.tree_all(lambda dim: dim is None, in_dims): 
     with interpreter.lower(): 
         return custom_function_call(autograd_function, *operands) 
 ```

Reply:
that should be
```
if pytree.tree_all(lambda dim: dim is None, in_dims):
        with interpreter.lower():
            return op(*operands)
```
the idea is that if none of the inputs are batched at the current level, then we just want to skip vmap and go to the next thing.